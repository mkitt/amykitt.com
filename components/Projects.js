// @flow
import React from 'react'
import { createSelector } from 'reselect'
import css, { media3, media4, media5 } from '../styles/css'
import View from './View'
import ProjectTile from './ProjectTile'
import type { Project } from '../types/app.js.flow'

type Props = {
  allowableWidth: number,
  columnCount: number,
  onClick: Function,
  projects: Array<Project>,
}

const wrapperStyles = css({
  marginLeft: -10,
})

const columnStyles = css(
  {
    display: 'inline-block',
    verticalAlign: 'top',
    width: 'calc(50% - 10px)',
    marginLeft: 10,
    overflow: 'hidden',
  },
  media3({ width: 'calc(33.33333% - 10px)' }),
  media4({ width: 'calc(25% - 10px)' }),
  media5({ width: 'calc(20% - 10px)' }),
)

const selectProjects = props => props.projects
const selectColumnCount = props => props.columnCount

const selectProjectsInColumns = createSelector(
  [selectProjects, selectColumnCount], (projects, columnCount) => {
    const columns = []
    for (let i = 0; i < columnCount; i += 1) { columns.push([]) }
    projects.forEach((item, index) => columns[index % columnCount].push(item))
    return columns
  },
)

export default class extends React.PureComponent {
  props: Props

  render() {
    const { allowableWidth, columnCount, projects, onClick } = this.props
    if (!projects || !projects.length) {
      return (
        <View>
          <p>Loading...</p>
        </View>
      )
    }
    const columns = selectProjectsInColumns({ projects, columnCount })
    return (
      <View className={wrapperStyles}>
        {columns.map((columnProjects, columnIndex) => (
          <View className={`Column ${columnStyles}`} key={`Column_${columnIndex + 1}`}>
            { columnProjects.map(project => (
              <ProjectTile
                allowableWidth={allowableWidth}
                onClick={onClick}
                key={`Project_${project.title}`}
                project={project}
              />
            ))}
          </View>
        ))}
      </View>
    )
  }
}
