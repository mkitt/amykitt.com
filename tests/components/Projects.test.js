import React from 'react'
import renderer from 'react-test-renderer'
import Projects from '../../components/Projects'

const noop = () => {}

describe('Projects', () => {
  let component

  beforeEach(() => {
    const projects = [1, 2, 3].map(num => (
      {
        body: `Project ${num} body`,
        id: `project-${num}`,
        tile: `project-${num}-tiles`,
        tileWidth: 600 * num,
        tileHeight: 400 * num,
        title: `Project ${num} title`,
      }
    ))

    component = (
      <Projects
        allowableWidth={1440}
        columnCount={4}
        onClick={noop}
        projects={projects}
      />
    )
  })

  it('shows the correct snapshot for the Projects component tree', () => {
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
