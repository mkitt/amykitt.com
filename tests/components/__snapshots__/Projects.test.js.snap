// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects shows the correct snapshot for the Projects component tree 1`] = `
<div
  className={
    Object {
      "data-css-19tn1yq": "",
    }
  }
>
  <div
    className="Column css-1pf4kuf"
  >
    <button
      className="css-1fqszn3 css-10xwomj"
      data-id="project-1"
      onClick={[Function]}
      title="Project 1 title"
      type="button"
    >
      <img
        alt="project 1-tiles"
        className=" css-bagnb1"
        height={400}
        src="project-1-tiles"
        width={600}
      />
    </button>
  </div>
  <div
    className="Column css-1pf4kuf"
  >
    <button
      className="css-1fqszn3 css-10xwomj"
      data-id="project-2"
      onClick={[Function]}
      title="Project 2 title"
      type="button"
    >
      <img
        alt="project 2-tiles"
        className=" css-bagnb1"
        height={800}
        src="project-2-tiles"
        width={1200}
      />
    </button>
  </div>
  <div
    className="Column css-1pf4kuf"
  >
    <button
      className="css-1fqszn3 css-10xwomj"
      data-id="project-3"
      onClick={[Function]}
      title="Project 3 title"
      type="button"
    >
      <img
        alt="project 3-tiles"
        className=" css-bagnb1"
        height={960}
        src="project-3-tiles"
        width={1440}
      />
    </button>
  </div>
  <div
    className="Column css-1pf4kuf"
  />
</div>
`;
